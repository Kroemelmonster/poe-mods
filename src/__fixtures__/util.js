export const findBy = key => (haystack, needle) =>
  haystack.find(({ [key]: value }) => value === needle);

export const findByPrimary = findBy('primary');
