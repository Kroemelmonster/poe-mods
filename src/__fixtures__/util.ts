import * as fs from 'fs';
import * as path from 'path';

import {
  AtlasNodeProps,
  BaseItemTypeProps,
  CraftingBenchOptionsProps,
  ModProps,
} from '../schema';

import AtlasNode from '../containers/AtlasNode';
import Item from '../containers/item/Item';
import MasterBenchOption from '../generators/MasterBenchOption';
import Mod from '../mods/Mod';
import PropsTable, { TableProps } from '../helpers/PropsTable';
import { Buildable } from '../interfaces/Buildable';

const json_cache: Partial<Tables> = {};

export const createTable = <P extends TableProps, T>(
  file: keyof Tables,
  constructor: Buildable<P, T>,
) => {
  if (json_cache[file] == null) {
    const body = fs.readFileSync(path.join(__dirname, `./${file}.json`));

    const json = JSON.parse(body.toString());

    // @ts-ignore
    json_cache[file] = new PropsTable(json, constructor);
  }

  return json_cache[file];
};

interface Tables {
  atlas: PropsTable<AtlasNodeProps, AtlasNode>;
  craftingbenchoptions: PropsTable<
    CraftingBenchOptionsProps,
    MasterBenchOption
  >;
  items: PropsTable<BaseItemTypeProps, Item>;
  mods: PropsTable<ModProps, Mod>;
}

export const createTables = () => {
  return {
    atlas: createTable('atlas', AtlasNode),
    craftingbenchoptions: createTable(
      'craftingbenchoptions',
      MasterBenchOption,
    ),
    items: createTable('items', Item),
    mods: createTable('mods', Mod),
  } as Tables;
};

export const byWorldAreaId = (id: string) => (node: AtlasNodeProps) =>
  node.world_area.id === id;

export const fromWorldAreaId = (
  id: string,
  atlas: PropsTable<AtlasNodeProps, AtlasNode>,
): AtlasNode => {
  return atlas.from(node => node.world_area.id === id);
};
