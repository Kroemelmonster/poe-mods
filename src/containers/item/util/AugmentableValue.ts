/**
 * a numeric value that was augmented in some way
 *
 * {augmented} is true if the the base value was changed to {value}
 */
export default interface AugmentableValue {
  value: number | [number, number];
  augmented: boolean;
}
