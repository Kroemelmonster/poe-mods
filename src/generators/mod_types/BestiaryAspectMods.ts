import ViewOnlyOrb from '../ViewOnlyOrb';
import Item from '../../containers/item';
import { Mod } from '../../mods';
import { ModProps } from '../../schema';

/**
 * Generator for all mods that are possible of the item is an Elder item
 */
export default class BestiaryAspectMods extends ViewOnlyOrb {
  /**
   * list of mods that can no longer appear in master vendor
   */
  public static readonly ASPECT_MODS: Set<Mod['props']['id']> = new Set([
    'GrantsCatAspect1',
    'GrantsBirdAspect1_',
    'GrantsSpiderAspect1',
    'GrantsCrabAspect1_',
  ]);

  public static build(mods: ModProps[]): BestiaryAspectMods {
    return new BestiaryAspectMods(
      mods
        .filter(BestiaryAspectMods.isBestiaryAspect)
        .map(props => new Mod(props)),
    );
  }

  /**
   * 
   * @param mod 
   * @returns true if the mod can only crafted with the Blood Altar from a
   *          Spirit Beast
   */
  public static isBestiaryAspect(mod: ModProps) {
    return BestiaryAspectMods.ASPECT_MODS.has(mod.id);
  }

  public modsFor(item: Item, whitelist: string[] = []) {
    // omit everything spawnable related. crafting is deterministic
    return super
      .modsFor(item, whitelist.concat('no_matching_tags', 'spawnweight_zero'))
      .map(({ spawnable, spawnweight, ...details }) => details);
  }
}
