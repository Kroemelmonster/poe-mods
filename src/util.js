// @flow
export * from './util/index';
