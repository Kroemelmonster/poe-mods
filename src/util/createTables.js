// @flow
import type {
  AtlasNodeProps,
  BaseItemTypeProps,
  CraftingBenchOptionsProps,
  ModProps,
} from '../schema';

import { AtlasNode, Item } from '../containers';
import { MasterBenchOption } from '../generators';
import { Buildable } from '../interfaces';
import { Mod } from '../mods';
import PropsTable, { type PropsWithPrimary } from './PropsTable';

function createTable<P: PropsWithPrimary>(
  props: P[],
  constructor: Class<Buildable<P>>,
) {
  return new PropsTable(props, constructor);
}

export const createAtlas = (props: AtlasNodeProps[]) => {
  return createTable(props, AtlasNode);
};

export const createItems = (props: BaseItemTypeProps[]) => {
  return createTable(props, Item);
};

export const createCraftingBenchOptions = (
  props: CraftingBenchOptionsProps[],
) => {
  return createTable(props, MasterBenchOption);
};

export const createMods = (props: ModProps[]) => {
  return createTable(props, Mod);
};
