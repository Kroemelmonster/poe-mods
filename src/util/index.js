// @flow
import { type Flags } from './Flags';
import { type ValueRange } from './ValueRange';

export type { Flags };
export type { ValueRange };
export { anySet } from './Flags';
export { default as Atlas } from './Atlas';
export { default as MasterBench } from './MasterBench';
export { default as Stat } from './Stat';
