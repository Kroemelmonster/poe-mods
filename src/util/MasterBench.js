// @flow
import type { Item } from '../containers';
import { MasterBenchOption } from '../generators';
import type { CraftingBenchOptionsProps } from '../schema';

import { anySet } from '../util/Flags';
import { MasterMod } from '../mods';

type OptionId = number;

/**
 */
export default class MasterBench {
  static build(props: CraftingBenchOptionsProps[], master_primary: ?number) {
    const options = props.filter(
      ({ npc_master_key }) =>
        // master_primary != null implies master_primary === npc_master_key
        master_primary == null || npc_master_key === master_primary,
    );

    if (options.length <= 0) {
      throw new Error(`no options found for '${String(master_primary)}'`);
    }

    // TODO allow customactions
    return new MasterBench(
      options
        .filter(({ mod }) => mod != null)
        .map(option => new MasterBenchOption(option)),
    );
  }

  +options: MasterBenchOption[];

  constructor(options: MasterBenchOption[]) {
    (this: any).options = options;
  }

  applyOptionTo(item: Item, option_id: OptionId): Item {
    const option = this.options.find(
      other => other.props.primary === option_id,
    );

    if (option === undefined) {
      throw new Error(`option '${option_id}' not found`);
    }

    return option.applyTo(item);
  }

  getAvailableMods(): MasterMod[] {
    return this.options.map(({ mod }) => mod).filter(Boolean);
  }

  /**
   * greps mod::applicableTo 
   */
  modsFor(item: Item, whitelist: string[] = []) {
    return this.options.reduce(
      (mods, option) => mods.concat(option.modsFor(item, whitelist)),
      [],
    );
  }
}
