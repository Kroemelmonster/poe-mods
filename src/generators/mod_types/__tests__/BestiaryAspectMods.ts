import { createTables } from '../../../__fixtures__/util';
import BestiaryAspectMods from '../BestiaryAspectMods';

const { items, mods } = createTables();

it('ignores internal spawnweights', () => {
  const item = items.fromName('Iron Greaves');
  const generator = BestiaryAspectMods.build(mods.all());

  expect(generator.modsFor(item).length).toBeGreaterThan(0);
  expect(
    generator
      .modsFor(item, ['domain_full'])
      .map(({ mod, ...details }) => ({ mod: mod.props.id, ...details })),
  ).toMatchSnapshot();
});
