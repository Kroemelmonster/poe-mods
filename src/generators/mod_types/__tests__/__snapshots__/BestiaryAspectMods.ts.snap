// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ignores internal spawnweights 1`] = `
Array [
  Object {
    "applicable": Object {
      "above_lld_level": false,
      "already_present": false,
      "domain_full": false,
      "lower_ilvl": false,
      "wrong_domain": false,
    },
    "mod": "GrantsCatAspect1",
  },
  Object {
    "applicable": Object {
      "above_lld_level": false,
      "already_present": false,
      "domain_full": false,
      "lower_ilvl": false,
      "wrong_domain": false,
    },
    "mod": "GrantsBirdAspect1_",
  },
  Object {
    "applicable": Object {
      "above_lld_level": false,
      "already_present": false,
      "domain_full": false,
      "lower_ilvl": false,
      "wrong_domain": false,
    },
    "mod": "GrantsSpiderAspect1",
  },
  Object {
    "applicable": Object {
      "above_lld_level": false,
      "already_present": false,
      "domain_full": false,
      "lower_ilvl": false,
      "wrong_domain": false,
    },
    "mod": "GrantsCrabAspect1_",
  },
]
`;
